src/readability/opportunities/handler.js

```diff
@@ -0,0 +1,227 @@
+import {
+  TOP_PAGES_LIMIT,
+} from '../shared/constants.js';
+
+const { AUDIT_STEP_DESTINATIONS } = Audit;
+// Use existing audit type or create new one if needed
+const AUDIT_TYPE_READABILITY = Audit.AUDIT_TYPES.READABILITY || 'readability';
+
+export async function processImportStep(context) {
+  const { site, finalUrl } = context;
+
+  const s3BucketPath = `scrapes/${site.getId()}/`;
+
+  return {
+    auditResult: { status: 'preparing', finalUrl },
+    fullAuditRef: s3BucketPath,
+    type: 'top-pages',
+    siteId: site.getId(),
+    allowCache: true,
+  };
+}
+
+// First step: sends a message to the content scraper to get page content for readability analysis
+export async function scrapeReadabilityData(context) {
+  const {
+    site, log, finalUrl, env, dataAccess,
+  } = context;
+  const siteId = site.getId();
+  const bucketName = env.S3_SCRAPER_BUCKET_NAME;
+
+  if (!bucketName) {
+    const errorMsg = 'Missing S3 bucket configuration for readability audit';
+    log.error(`[ReadabilityProcessingError] ${errorMsg}`);
+    return {
+      status: 'PROCESSING_FAILED',
+      error: errorMsg,
+    };
+  }
+
+  log.info(`[ReadabilityAudit] Step 1: Preparing content scrape for readability audit for ${site.getBaseURL()} with siteId ${siteId}`);
+
+  // Get top pages for readability analysis
+  const { SiteTopPage } = dataAccess;
+  const topPages = await SiteTopPage.allBySiteIdAndSourceAndGeo(site.getId(), 'ahrefs', 'global');
+
+  log.info(`[ReadabilityAudit] Found ${topPages?.length || 0} top pages for site ${site.getBaseURL()}`);
+
+  if (!isNonEmptyArray(topPages)) {
+    log.info(`[ReadabilityAudit] No top pages found for site ${siteId} (${site.getBaseURL()}), skipping audit`);
+    return {
+      status: 'NO_OPPORTUNITIES',
+      message: 'No top pages found, skipping audit',
+    };
+  }
+
+  // Take top pages by traffic, sorted descending
+  const urlsToScrape = topPages
+    .map((page) => ({ url: page.getUrl(), traffic: page.getTraffic(), urlId: page.getId() }))
+    .sort((a, b) => b.traffic - a.traffic)
+    .slice(0, TOP_PAGES_LIMIT);
+
+  log.info(`[ReadabilityAudit] Top ${TOP_PAGES_LIMIT} pages for site ${siteId} (${site.getBaseURL()}): ${JSON.stringify(urlsToScrape, null, 2)}`);
+
+  return {
+    auditResult: {
+      status: 'SCRAPING_REQUESTED',
+      message: 'Content scraping for readability audit initiated.',
+      scrapedUrls: urlsToScrape,
+    },
+    fullAuditRef: finalUrl,
+    // Data for the CONTENT_SCRAPER
+    urls: urlsToScrape,
+    siteId,
+    jobId: siteId,
+    processingType: 'default',
+  };
+}
+
+// Second step: processes scraped data to create readability opportunities
+export async function processReadabilityOpportunities(context) {
+  const {
+    site, log, s3Client, env, audit,
+  } = context;
+  const siteId = site.getId();
+  const bucketName = env.S3_SCRAPER_BUCKET_NAME;
+
+  if (!bucketName) {
```

Comment:
If you check for this, you should also check for other statuses of prior steps and abort early, e.g. `NO_OPPORTUNITIES` or `PROCESSING_FAILED`.